test/upgrade/main.go

```diff
@@ -337,3 +375,96 @@ func createGameServerFile(agonesVersion string) string {
 
 	return gsPath
 }
+
+// Create a game server every five seconds until the context is cancelled. The game server container
+// be the same binary version as the game server file. The SDK version is always the same as the
+// version of the Agones controller that created it. The Game Server shuts itself down after the
+// tests have run as part of the `sdk-client-test` logic.
+func createGameServers(ctx context.Context, gsPath string) {
+	args := []string{"create", "-f", gsPath}
+	ticker := time.NewTicker(5 * time.Second)
+
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			_, err := runExecCommand(KubectlCmd, args...)
+			// TODO: Do not ignore error if unable to create due to something other than cluster scale up
+			if err != nil {
+				log.Printf("Could not create Gameserver %s: %s", gsPath, err)
+			}
+		}
+	}
+}
+
+// watchGameServerPods watches all game server pods for CrashLoopBackOff. Errors if the number of
+// CrashLoopBackOff backoff pods exceeds the number of acceptedFailures.
+func watchGameServerPods(kubeClient *kubernetes.Clientset, stopCh chan struct{}, failedPods map[string]podLog, acceptedFailures int) {
+	// Filter by label agones.dev/role=gameserver to only game server pods
+	labelOptions := informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+		opts.LabelSelector = "agones.dev/role=gameserver"
+	})
+	kubeInformerFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, 5*time.Second,
+		informers.WithNamespace("default"), labelOptions)
+	podInformer := kubeInformerFactory.Core().V1().Pods().Informer()
+
+	_, err := podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newObj interface{}) {
+			newPod := newObj.(*v1.Pod)
+			for _, cs := range newPod.Status.ContainerStatuses {
+				if cs.Name != "sdk-client-test" || cs.State.Waiting == nil || cs.State.Waiting.Reason != "CrashLoopBackOff" {
+					continue
+				}
+				gsVersion := newPod.Labels["agonesVersion"]
+				sdkVersion := newPod.Annotations["agones.dev/sdk-version"]
+				log.Printf("%s for pod: %s with game server binary version %s, and SDK version %s", cs.State.Waiting.Reason, newPod.Name, gsVersion, sdkVersion)
+				// Put failed pods into the map until it reaches capacity.
+				failedPods[newPod.Name] = podLog{GameServerVersion: gsVersion, SdkVersion: sdkVersion}
+				if len(failedPods) > acceptedFailures {
+					log.Fatalf("Too many Game Server pods in CrashLoopBackOff: %v", failedPods)
+				}
+			}
+		},
+	})
+	if err != nil {
+		log.Fatal("Not able to create AddEventHandler", err)
+	}
+
+	go podInformer.Run(stopCh)
```

Comment:
I still don't quite get the usage of the `stopCh` here. Seems no one is sending signal to stopCh, so this line is equal to `go podInformer.Run()`? If so when does this go routine get stopped?

Reply:
It's a required paramenter for the .Run() function. I think it's used instead of context.Context to stop the informer anywhere in the code. We don't use it since nothing we want the informer to run until the main thread completes.

Reply:
Oh I see, I thought it's optional :)